### PR TITLE
Merchant Fulfillment API MWS

### PR DESCRIPTION
**Pull request by cmchecha**

### CODE REVIEW

**Comment by GriceTurrble** [comment]
Code coverage issue is something I'm working on separately, so no need to worry about the coverage decrease in this PR. Post-0.9 we will have a better standard for coverage of new features.

**Comment by GriceTurrble** [comment]
@bobspadger, can you independently verify this code working in your production? As before, I have no seller account, so I cannot test it against MWS.

**Comment by Bobspadger** [comment]
@GriceTurrble I'll try, I need to read the docs and work out if my account is activated for this and that I won't cause our order process / customers any issues.

**Comment by Bobspadger** [comment]
Testing - got stuck:

```python
import mws
from credentials import creds
# Make some easier to read globals
SECRET_KEY = creds['amazon']['secret_key']
AWS_ACCESS_KEY = creds['amazon']['AWS_access_id']
SELLER_ID = creds['amazon']['sellerid']
MARKETPLACE_ID = creds['amazon']['marketplaceids']['UK']

shipping = mws.MerchantFulfillment(access_key=AWS_ACCESS_KEY, secret_key=SECRET_KEY, account_id=SELLER_ID, region='UK')



###
# Get service status
###

service_status = shipping.get_service_status()
print(service_status.parsed)

"""
Passes, status green good to carry on
"""


###
# Get eligible shipping services
###

"""
I'm using real order items and id's here

Note: Quantity : 1 fails, Quantity: '1' works and allows params to be encoded. Not sure if this is expected?

"""

item_list = [{'OrderItemId': 'REDACTED', 'Quantity': '1'}]
ship_fron = {
    'Name': 'Absolute Music',
    'AddressLine1': 'REDACTED',
    'AddressLine2': 'REDACTED',
    'City': 'REDACTED',
    'Email': 'REDACTED',
    'PostalCode': 'REDACTED',
    'Phone': 'REDACTED'
}
package = {'Length': '10',
           'Width': '10',
           'Height': '10',
           'Unit': 'Centimetres'
           }
weight = {'Value': '1',
          'Unit': 'g'}
ship_service = {'DeliveryExperience': 'DeliveryConfirmationWithSignature',
                'CarrierWillPickUp': 'true'}

services = shipping.get_eligible_shipping_services("XXX-XXXXXXX-XXXXXXX", item_list=item_list, ship_from_address=ship_fron,
                                                   package_dimensions=package, weight=weight, shipping_service_options=ship_service)
print(services.parsed)
```



and I get

```

<ErrorResponse xmlns="https://mws.amazonservices.com/MerchantFulfillment/2015-06-01">
  <Error>
    <Type>Sender</Type>
    <Code>MalformedInput</Code>
    <Message>Unexpected complex element termination</Message>
  </Error>
  <RequestId>9bcacad8-bd4d-495b-b863-1ffc86fb7c5e</RequestId>
</ErrorResponse>

```


Can't see anything immediately incorrect - do you have any info with regards to my inputs?? @cmchecha

**Comment by GriceTurrble** [comment]
Just noticed this PR was targeting `master`. Target changed to `develop` to coincide with our workflow.

File: mws/utils.py [hunk 1]
```diff
@@ -66,7 +66,7 @@ def _parse_node(self, node):
             key, val = self._namespace_split(key, ObjectDict({'value': val}))
             node_tree[key] = val
         # Save childrens
-        for child in node.getchildren():
+        for child in node:
```
Comment (GriceTurrble) [inline mws/utils.py]: This is a fairly basic upstream process that seems to have been working so far. Changing this now could have some nasty aftereffects without extensive testing.

@cmchecha , can you provide some detail here, please?

Comment (cmchecha) [reply mws/utils.py]: Output in my production:
DeprecationWarning: This method will be removed in future versions.  Use 'list(elem)' or iteration over elem instead.
  for child in node.getchildren():

But we can return to the previous version.

**Comment by cmchecha** [comment]
@Bobspadger I found the error, it was the list of items. 
Correction in your body:
Centimetres to centimeters in package and add ContryCode in your ship_from, this is required

```
    item_list = [{'OrderItemId': 'REDACTED', 'Quantity': '1'}]
        ship_fron = {
            'Name': 'REDACTED',
            'AddressLine1': 'REDACTED',
            'AddressLine2': 'REDACTED',
            'City': 'REDACTED',
            'Email': 'test@mail.com',
            'PostalCode': 'REDACTED',
            'Phone': 'REDACTED',
            'CountryCode': "REDACTED"
        }
        package = {'Length': '10',
                'Width': '10',
                'Height': '10',
                'Unit': 'centimeters'
                }
        weight = {'Value': '1',
                'Unit': 'g'}
        ship_service = {'DeliveryExperience': 'DeliveryConfirmationWithSignature',
                        'CarrierWillPickUp': 'true'}
```

File: mws/utils.py (hunk 1 above)
Comment (GriceTurrble) [reply mws/utils.py]: Ah yes, I see that now. Did some digging: `getchildren` is deprecated as of 2.7 ([doc](https://docs.python.org/2.7/library/xml.etree.elementtree.html#xml.etree.ElementTree.Element.getchildren)) and 3.2 ([doc](https://docs.python.org/3.6/library/xml.etree.elementtree.html#xml.etree.ElementTree.Element.getchildren)).

Spot-checked the change locally, and I confirm it works that same as original. We can accept that change. Good catch, thank you. :)

**Comment by Bobspadger** [comment]
Those last updates are providing me with better error codes, working through a test now

**Comment by Bobspadger** [comment]
```xml
<ErrorResponse xmlns="https://mws.amazonservices.com/MerchantFulfillment/2015-06-01">
  <Error>
    <Type>Sender</Type>
    <Code>TermsAndConditionsNotAccepted</Code>
  </Error>
  <RequestId>7d3d2cf5-6686-4a7e-8713-8791d824fc8b</RequestId>
</ErrorResponse>
```


Not sure if this is because I need to accept terms and conditions, documentation @amazon is lacking on this

**Comment by GriceTurrble** [comment]
Docs cover this somewhat:

http://docs.developer.amazonservices.com/en_UK/merch_fulfill/MerchFulfill_ErrorCodes.html

> The seller has not yet agreed to Amazon's or the carrier's terms and conditions. You can accept terms and condition in Seller Central ([Europe](https://sellercentral.amazon.co.uk/gp/shipping-manager/terms-and-conditions.html)) ([US](https://sellercentral.amazon.com/gp/shipping-manager/terms-and-conditions.html)).

**Comment by Bobspadger** [comment]
@cmchecha 

I found you updates provided me with those error messages to fix, so thanks, it was that list items. Its returning proper responses now

@GriceTurrble  Yep, I just found that too - being a bit slow today :(

I now have responses from Amazon, but not quite as they are giving me with their seller central.

I'm just trying to work out why this is at present but the communications aspect of it is looking good :)

**Comment by Bobspadger** [comment]
I have a support request open with Amazon on this.

To me, it looks like we are ok for the communications part of it (only tested on `GetEligibleShippingServices` 
However, the API is returning to me that everything is currently Temporarily unavailable.

Once I've got confirmation from Amazon that its actually them needing to activate the services on my account and doing so, I can test and confirm the rest of the label creation etc.

Looks good so far though.

**Comment by Bobspadger** [comment]
Just an update on this.

I now have the full merchant fulfilment API working and returning labels.

I'm happy this is fully working now.

**Pull request merged**